### ISSUE TITLE
Error with Flux.crossentropy

### ISSUE THREAD

**Issue by xiaodaigh** [FluxML/Flux.jl#435]
```
Flux.crossentropy([false, true, false, false], [1.0, 0.0, 0.0, 0.0] )
```
gives `NaN` or `Inf`

It should give `0`

**Comment by Paethon** [comment]
No it should not.

A crossentropy-loss of 0.0 would mean you have a perfect fit between the expected output `[false, true, false, false]` and the observed output `[1.0, 0.0, 0.0, 0.0]`, which is clearly not the case. 

In this case `Inf` could be considered the correct output (`Nan` is obviously bad). Usually frameworks clamp values of 0.0 to a very small epsilon to circumvent numerical instability (which the Flux-implementation does not), so maybe this should be changed ...

**Comment by boathit** [comment]
For the purpose of numerical stability, `crossentropy` should be discouraged and `logitcrossentropy` or `binarylogitcrossentropy` is strongly recommended.

**Comment by Paethon** [comment]
Is there a specific reason `crossentropy` is not implemented in a numerically stable way? Because I would think if it is in the framework it should be implemented "properly", regardless whether its use is discouraged or not.

**Comment by MikeInnes** [comment]
I agree and there's no specific reason I know of, so if you have any stability improvements they'd be very welcome.

**Comment by Paethon** [comment]
Ahh, by the way, the first argument is the predicted and the second is the observed result. So this should probably be 

```julia
Flux.crossentropy([1.0, 0.0, 0.0, 0.0], [false, true, false, false])
```

**Comment by opus111** [comment]
Hi Mike.  I'm new here, but a big fan of Julia and Flux.  Thank you for writing it :-D

I ran into the same stability issues with crossentropy() and wrote my own.  Here it is.  Use if you see fit

    const ε = eps(1.0)
    crossentropy(x) = y > 0.5 ? -(log(ŷ+ε)) : -(log((1.0-ŷ)+ε))
